### PR TITLE
Remove assumption that Azure security group is hard coded by Juju

### PR DESCRIPTION
**Pull request by wallyworld**
## Description of change

When specifying a bespoke virtual network for a Juju model on azure, use the security group (if any) associated with the user specified subnets instead of Juju creating one with a hard coded name.
Juju will now only create network resources - virtual networks, subnets, security groups - for models where there is no user specified network model-config.
When opening/closing ports on an instance, Juju will look up the security groups to use off the instances's NIC(s) instead of assuming a hard coded one. Followup work needed to consider the new endpoint binding modelling.

## QA steps

bootstrap on azure
juju add-model and ensure deploy mysql and expose work as before with the hard coded artefacts
create a bespoke resource group (test-rg) and virtual network (test-vn) with subnets and a security group
juju add-model --config network=test-rg/test-vn
juju add-space to create a space with one of the subnets
juju deploy mysql --bind db=myspace --constraints="spaces=alpha"
juju expose mysql
check that the model's resource group has the machine and NICs, and that these use the subnets in the bespoke vnet
check the bespoke security group for the mysql port rule
juju show-unit mysql/0 and check the network info matches the bound space subnet

### CODE REVIEW

File: provider/azure/networking.go [hunk 1]
```diff
@@ -216,26 +215,3 @@ func nextSecurityRulePriority(group network.SecurityGroup, min, max int32) (int3
 		"no priorities available in the range [%d, %d]", min, max,
 	)
 }
-
```
Comment (wallyworld) [inline provider/azure/networking.go:219]: moved to upgrades.go below as this is only used there and want to keep networking.go free of such upgrade related methods

File: provider/azure/networking.go [hunk 2]
```diff
@@ -111,34 +110,11 @@ type newSecurityRuleParams struct {
...
-			// Two rules cannot have the same name.
-			name:        fmt.Sprintf("%s%d", apiSecurityRulePrefix, apiPort),
-			description: "Allow API connections to controller machines",
-			destPrefix:  controllerSubnetPrefix,
-			port:        apiPort,
-			// Two rules cannot have the same priority and direction.
-			priority: securityRuleInternalAPIInbound + i,
-		}))
-	}
-	securityRules = append(securityRules, extraRules...)
```
Comment (wallyworld) [inline provider/azure/networking.go:140]: extracted to a new method networkSecurityRules

**Comment by wallyworld** [comment]
[azure subnet security group modelling](https://trello.com/c/t3GG1km7/1169-azure-subnet-security-group-modelling)

File: provider/azure/instance.go [hunk 3]
```diff
@@ -184,50 +182,110 @@ func (inst *azureInstance) Addresses(ctx context.ProviderCallContext) (corenetwo
...
+		privateIpAddress := ipConfiguration.PrivateIPAddress
+		if privateIpAddress == nil {
+			continue
+		}
+		securityGroup := nic.NetworkSecurityGroup
+		if securityGroup == nil {
+			idParts := strings.Split(to.String(ipConfiguration.Subnet.ID), "/")
+			lenParts := len(idParts)
+			subnetClient := network.SubnetsClient{BaseClient: env.network}
+			subnet, err := subnetClient.Get(ctx, idParts[lenParts-7], idParts[lenParts-3], idParts[lenParts-1], "networkSecurityGroup")
```
Comment (ycliuhw) [inline provider/azure/instance.go:214]: is there a builtin parser for this?

**Review by ycliuhw — APPROVED** [review]
LGTM, thanks

File: provider/azure/instance.go (hunk 3 above)
Comment (wallyworld) [reply provider/azure/instance.go:214]: I wish there was, I couldn't find one and other example code seemed to split it up manually.

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**